src/get-ipfs.ts

```diff
@@ -21,125 +26,144 @@ interface IpfsPkg extends ipfs {
-    } else {
-      return false
-    }
+    return !!(id.id && id.agentVersion)
   } catch (err) {
     return false
   }
 }
 
-export const loadWindowIpfs = async (config: config): Promise<ipfs | null> => {
-  let ipfs = (window as IpfsWindow).ipfs
-  if (ipfs) {
-    if (ipfs.enable) {
-      const permissions = normalizePermissions(config.permissions)
-      ipfs = await ipfs.enable({
-        commands: permissions
-      })
-    }
-    const isWorking = await ipfsIsWorking(ipfs)
-    if (isWorking) {
-      return ipfs as ipfs
-    } else {
-      return null
-    }
-  }
-  return null
-}
 
-export const loadIpfsScript = (): Promise<void> => {
-  return new Promise((resolve, reject) => {
-    const script = document.createElement('script')
-    script.src = 'https://unpkg.com/ipfs@0.38.0/dist/index.min.js'
-    script.onload = async () => {
-      return resolve()
-    }
-    script.onerror = reject
-    document.body.appendChild(script)
-  })
+export async function loadJsIpfs(config: config): Promise<ipfs | null> {
+  const ipfsPkg = await (() => { switch (typeof config.jsIpfs) {
+    case "function":  return config.jsIpfs()
+    case "string":    return loadJsIpfsThroughScript(config.jsIpfs)
+    default:          return loadJsIpfsThroughScript(UNPKG_URL)
+  }})()
+
+  if (!ipfsPkg?.create) return null
+  const peers = (config.peers || []).concat(config.browserPeers || [])
+  const ipfs = await ipfsPkg.create({ config: { Addresses: { Swarm: peers }}})
+  return await ipfsIsWorking(ipfs) ? ipfs : null
 }
 
-export const loadJsIpfs = async (config: config): Promise<ipfs | null> => {
-  try {
-    await loadIpfsScript()
-  } catch (err) {
-    return null
-  }
 
-  const ipfsPkg = (window as IpfsWindow).Ipfs
-  if (!ipfsPkg || !ipfsPkg.create) {
-    return null
-  }
-  const ipfs = await ipfsPkg.create()
-  const isWorking = await ipfsIsWorking(ipfs)
+export async function loadWindowIpfs(config: config): Promise<ipfs | null> {
+  let ipfs = (window as IpfsWindow).ipfs
+  if (!ipfs) return null
 
-  if (isWorking) {
-    return ipfs
-  } else {
-    return null
+  if (ipfs.enable) {
+    const permissions = normalizePermissions(config.permissions)
+    ipfs = await ipfs.enable({ commands: permissions })
   }
+
+  return await ipfsIsWorking(ipfs)
+    ? ipfs as ipfs
+    : null
 }
 
-const getIpfs = async (config: config = {}): Promise<ipfs> => {
-  // if already instantiated
-  if (ipfsInstance) {
-    return ipfsInstance
-  }
 
-  let peers = config.peers || []
-  ipfsInstance = await loadWindowIpfs(config)
-  if (ipfsInstance) {
-    console.log('window.ipfs is available!')
-    peers = peers.concat(config.localPeers || [])
-  } else {
-    console.log('window.ipfs is not available, downloading from CDN...')
-    ipfsInstance = await loadJsIpfs(config)
-    peers = peers.concat(config.browserPeers || [])
-  }
 
+// ㊙️
+
+
+async function connectPeers(ipfs: ipfs, peers: string[] = []): Promise<ipfs> {
   if (!ipfsInstance) {
```

Comment:
I think we want to this be up in `getIpfs()` instead.

This function doesn't use `ipfsInstance` it just uses whichever `ipfs` is passed to it.

There is a question of if we should throw an error or return `null` if we can't instantiate any ipfs. I went with "throwing an error" because the point of this package is to _always_ get an IPFS instance in the developers hands.

Reply:
Whoops, global variables got me 😬 
Will put this back.

Yeah, I don't know with javascript.
In other languages I'd prefer a `Maybe` (ie. null in this case).
But I agree regarding the purpose of this library.
So I guess throwing an error is fine for now.